metabase/metabase
# Pass parameter values as JSON

Fixes https://github.com/metabase/metabase/issues/46240

The problem was that we pass parameter values as separate query string params in embeded cards & dashboards. It looks like this `?param1=A&param2=10&param3=true`. The issue is that we cannot correctly parse these parameters on the BE since any type information was lost during the parameter values serialization. Is `10` a number or a string? Is `true` a boolean or a string? We don't know. We could do either 1) pass parameters from the FE in another way where this information is not lost OR 2) find the corresponding fields for each parameter, check their types, and convert the values correctly. The latter is difficult and could be a performance concern, so this PR implements 1).

Now parameter values are passed as `?parameters=<json>`. This is exactly the same as public dashboards work right now. To maintain backward compatibility we still accept separate parameter values if `parameters` is not present or when it cannot be parsed. I thought that breaking the API contract for `query` endpoints is not acceptable in a patch release.

How to verify:
- Enable embedding in Admin
- New -> Question -> Products -> Save as `Q1`
- New -> Dashboard -> Add `Q1` -> Add Text filter for `Product.Category` -> Set the parameter to be just `Input` -> Save
- Share -> Embed -> Static -> Enable the parameter -> Preview -> Look and feel
- Try to enter a numeric value in the parameter widget
- The query should not fail. There should be no results.
- We have e2e tests that check embedded cards & dashboards and downloads in them.
